src/devices/stereoCamera.cpp

```diff
@@ -1472,32 +1327,59 @@ bool StereoCamera::setImageROIoffsetYEmu(int offsetY) {
     //std::cout << "Setting Image ROI offsetY=" << std::to_string(offsetY) << std::endl;
     bool success = false;
 
-    if(cameras[0].IsGrabbing())
-        cameras[0].StopGrabbing();
-    if(cameras[1].IsGrabbing())
-        cameras[1].StopGrabbing();
+    if(cameras.IsGrabbing())
+        stopGrabbing();
 
-    // ace Classic/U/L GigE Cameras
-   // int maxWidth = cameras[0].Width.GetMax();
-   // int maxHeight = cameras[0].Height.GetMax();
-    // other cameras
     int maxHeight = getImageROIheightMax();
     int height = getImageROIheight();
 
     if(maxHeight - offsetY < 16)
         offsetY = maxHeight - 16;
-    //if(height + offsetY > maxHeight)
-    //    return;
-
-    int modVal=offsetY%4;
+    int modVal=offsetY%16;
     if(modVal != 0)
         offsetY -= modVal;
 
     if (height + offsetY <= maxHeight && cameras[0].OffsetY.IsWritable() && cameras[1].OffsetY.IsWritable() ) {
         success = cameras[0].OffsetY.TrySetValue(offsetY) &&
             cameras[1].OffsetY.TrySetValue(offsetY);
-    } 
-    cameras[0].StartGrabbing(GrabStrategy_OneByOne, GrabLoop_ProvidedByInstantCamera);
-    cameras[1].StartGrabbing(GrabStrategy_OneByOne, GrabLoop_ProvidedByInstantCamera);
+    }
+    startGrabbing();
     return success;
-}
\ No newline at end of file
+}
+
+void StereoCamera::safelyCloseCameras() {
+    if(!cameras.IsOpen())
+        return;
+
+    if(cameras.IsGrabbing())
+        cameras.StopGrabbing();
+    cameras.Close();
+
+    if(cameraImageEventHandler) {
```

Comment:
Event Handlers registered with delete_cleanup cleanup procedure are automatically cleaned up after deregistering, so there should be no need for manual deletion (cameraImageEventHandler->DestroyImageEventHandler). 

https://zh.docs.baslerweb.com/pylonapi/cpp/class_pylon_1_1_c_instant_camera#function-deregistercameraeventhandler

Also, they should all be set to nullptr (see cameraConfigurationEventHandler1) or none of them should be set to nullptr. I recommend not setting them, as they are automatically deleted upon deregistering, and only really used when they are being created.

Same applies to Hardware Trigger Configurations.

https://zh.docs.baslerweb.com/pylonapi/cpp/class_pylon_1_1_c_instant_camera#function-registerconfiguration

Reply:
All are now set to nullptr. Also added the same object management in single camera class, where software/hardware trigger configuration instances were not deregistered properly upon configuration change and camera closing.